File: 2025/YoungjoonPark/week9/2025-08-02-gas-station.py
# KDD LeetCode Study Week 9: Greedy
# Assignment 1
# Author: Youngjoon Park
# URL: https://leetcode.com/problems/gas-station

from typing import List

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        n = len(gas)
        total_gas = sum(gas)
        total_cost = sum(cost)
        
        if total_gas < total_cost:
            return -1
        
        current_tank = 0
        start_station = 0
        
        for i in range(n):
            current_tank += gas[i] - cost[i]
            
            if current_tank < 0:
                start_station = i + 1
                current_tank = 0
        
        return start_station

File: 2025/SoyoungJung/week8/2025-08-02-Gas-Station.ts
function canCompleteCircuit(gas: number[], cost: number[]): number {
  let tank = 0;
  let totalGas = 0;
  let totalCost = 0;
  let startIndex = 0;

  for (let i = 0; i < gas.length; i++) {
    totalGas += gas[i];
    totalCost += cost[i];
    tank += gas[i] - cost[i];

    if (tank < 0) {
      startIndex = i + 1;
      tank = 0;
    }
  }

  return totalGas >= totalCost ? startIndex : -1;
}
